### PR TITLE
Bump up the Microsoft.CodeAnalysis version used by packages in the repo

### LINKED ISSUE #6089: Revise usages of `IObjectCreationOperation.Constructor`
`IObjectCreationOperation.Constructor` can be null in VB for this error scenario:

```vb
Class C
    Protected Structure S
    End Structure
End Class

Class D
    Shared Sub M(o)
        M(New C.S())
    End Sub
End Class
```

Revise all analyzers that dereferences `IObjectCreationOperation.Constructor`, add test that causes NRE, then fix it.

### PR DESCRIPTION
**Pull request by mavasani**
Bump up the Microsoft.CodeAnalysis version used by packages in the repo to 3.11.0 (VS2019 16.11), except PublicApiAnalyzers,  and also bump up the version of the generated analyzer packages in the repo.

This brings in the nullable annotated API surfaces for ISymbol, IOperation and syntax APIs, which generates a large number of new nullable warnings. This PR also fixes/suppresses these warnings as appropriate.

Closes #6089

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs [hunk 1]
```diff
@@ -287,7 +287,7 @@ void AnalyzeFixerWithFixAll(INamedTypeSymbol fixer, SymbolAnalysisContext contex
                         {
                             if (IsViolatingCodeActionObjectCreation(objectCreation))
                             {
-                                Diagnostic diagnostic = objectCreation.CreateDiagnostic(OverrideCodeActionEquivalenceKeyRule, objectCreation.Constructor.ContainingType, EquivalenceKeyPropertyName);
+                                Diagnostic diagnostic = objectCreation.CreateDiagnostic(OverrideCodeActionEquivalenceKeyRule, objectCreation.Constructor!.ContainingType, EquivalenceKeyPropertyName);
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs:290]: Is there a check that eliminates the possibility of https://github.com/dotnet/roslyn-analyzers/issues/6089?

Comment (Youssef1313) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs:290]: Oh it's below.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CancellationTokenParametersMustComeLast.cs [hunk 2]
```diff
@@ -90,7 +90,7 @@ public override void Initialize(AnalysisContext context)
                     // Ignore parameters that have any of these attributes.
                     // C# reserved attributes: https://learn.microsoft.com/dotnet/csharp/language-reference/attributes/caller-information
                     while (last >= 0
-                        && HasCallerInformationAttribute(methodSymbol.Parameters[last], callerInformationAttributes))
+                        && methodSymbol.Parameters[last].HasAnyAttribute(callerInformationAttributes))
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/CancellationTokenParametersMustComeLast.cs:93]: Nice!

**Review by Youssef1313 — APPROVED** [review]
Haven't looked into every line (it's a quite large PR), but going through quickly, LGTM!

File: eng/Versions.props [hunk 3]
```diff
@@ -60,17 +60,17 @@
     -->
     <!-- Microsoft.CodeAnalysis versions for different purposes. -->
     <!-- Surface area that various projects compile against. These should have source-build reference packages -->
-    <MicrosoftCodeAnalysisForRoslynDiagnosticsAnalyzersVersion>3.7.0</MicrosoftCodeAnalysisForRoslynDiagnosticsAnalyzersVersion>
+    <MicrosoftCodeAnalysisForRoslynDiagnosticsAnalyzersVersion>3.11</MicrosoftCodeAnalysisForRoslynDiagnosticsAnalyzersVersion>
```
[the remark this answers is not in the document]
Comment (mavasani) [reply eng/Versions.props]: Thanks, changed.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotDisableRequestValidation.cs [hunk 4]
```diff
@@ -78,7 +77,8 @@ public override void Initialize(AnalysisContext context)
 
                             if (constructorArguments.Length == 1 &&
                                 constructorArguments[0].Kind == TypedConstantKind.Primitive &&
-                                constructorArguments[0].Value.Equals(false))
+                                constructorArguments[0].Value is bool value &&
+                                !value)
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/DoNotDisableRequestValidation.cs]: Done, thanks.

File: eng/Versions.props [hunk 5]
```diff
@@ -60,17 +60,17 @@
...
+    <MicrosoftCodeAnalysisVersionForPerfSensitiveAnalyzersTests>3.11.0</MicrosoftCodeAnalysisVersionForPerfSensitiveAnalyzersTests>
     <MicrosoftCodeAnalysisVersionForResxSourceGenerators>4.0.1</MicrosoftCodeAnalysisVersionForResxSourceGenerators>
-    <MicrosoftCodeAnalysisVersionForNetAnalyzers>3.3.1</MicrosoftCodeAnalysisVersionForNetAnalyzers>
-    <MicrosoftCodeAnalysisVersionForTextAnalyzers>3.3.1</MicrosoftCodeAnalysisVersionForTextAnalyzers>
-    <MicrosoftCodeAnalysisVersionForCodeAnalysisAnalyzers>3.3.1</MicrosoftCodeAnalysisVersionForCodeAnalysisAnalyzers>
-    <MicrosoftCodeAnalysisVersionForToolsAndUtilities>3.3.1</MicrosoftCodeAnalysisVersionForToolsAndUtilities>
+    <MicrosoftCodeAnalysisVersionForNetAnalyzers>3.11.0</MicrosoftCodeAnalysisVersionForNetAnalyzers>
+    <MicrosoftCodeAnalysisVersionForTextAnalyzers>3.11.0</MicrosoftCodeAnalysisVersionForTextAnalyzers>
+    <MicrosoftCodeAnalysisVersionForCodeAnalysisAnalyzers>3.11.0</MicrosoftCodeAnalysisVersionForCodeAnalysisAnalyzers>
+    <MicrosoftCodeAnalysisVersionForToolsAndUtilities>3.11.0</MicrosoftCodeAnalysisVersionForToolsAndUtilities>
```
Comment (mavasani) [inline eng/Versions.props:73]: @mmitche @MichaelSimons This PR moves the referenced `Microsoft.CodeAnalysis` versions for the packages to `3.11.0` (shipped with VS2019 16.11). However, this fails the pre-build detection step for source build that you added with #6610. I see that https://github.com/dotnet/source-build-reference-packages/pull/692 added the SB reference packages for the prior versions of the dependency packages, and are being mapped from here: https://github.com/dotnet/roslyn-analyzers/blob/b4dc141b0568be5aa135965d30c49c740b706be1/eng/Version.Details.xml#L8-L12

I presume this PR would require an update to the SB packages to add the `3.11.0` version of the packages before we can pass the pre-build detection leg. Is that correct? If so, who can help me out to perform this update in https://github.com/dotnet/source-build-reference-packages? If there are documentation pointers on how to go about this, please share them so I can try to do this myself. Thanks!

Comment (mavasani) [reply eng/Versions.props:73]: Ah, I see the steps at https://github.com/dotnet/source-build-reference-packages/tree/main#adding-new-packages. I'll give it a shot.

Comment (mavasani) [reply eng/Versions.props:73]: I tried executing those steps locally, but I get a build failure at the first step itself:

> generate.cmd -package microsoft.codeanalysis,3.11.0

**Error:**
```
  PackageSourceGenerator -> Generating package source for Humanizer.Core (v2.2.0)...
  Determining projects to restore...
  Restored C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj (in 5 ms).
  Restored C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGeneratorTask\PackageSourceGenera
  torTask.csproj (in 136 ms).
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018: The "G
etPackageItems" task failed unexpectedly.
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018: System
.ArgumentException: Found strong name key that doesn't map: Key=979442b78dfc278e, Assembly=Humanizer
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018:    at
Microsoft.DotNet.SourceBuild.Tasks.GetPackageItems.TryGetStrongNameData(AssemblyName assemblyName, StrongNameData& stro
ngNameData) in C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGeneratorTask\GetPackageItems
.cs:line 241
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018:    at
Microsoft.DotNet.SourceBuild.Tasks.GetPackageItems.SetCompileItems(PackageArchiveReader packageArchiveReader, TargetFra
meworkRegexFilter targetFrameworkRegexFilter) in C:\source-build-reference-packages\src\packageSourceGenerator\PackageS
ourceGeneratorTask\GetPackageItems.cs:line 153
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018:    at
Microsoft.DotNet.SourceBuild.Tasks.GetPackageItems.Execute() in C:\source-build-reference-packages\src\packageSourceGen
erator\PackageSourceGeneratorTask\GetPackageItems.cs:line 84
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018:    at
Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
C:\source-build-reference-packages\src\packageSourceGenerator\PackageSourceGenerator.proj(106,5): error MSB4018:    at
Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext ta
skLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

**Warnings:**
```
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'SQLitePCLRaw.core.dll' in any of the provided search d
irectories. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.csharp.work
spaces\3.11.0\microsoft.codeanalysis.csharp.workspaces.3.11.0.csproj::TargetFramework=netstandard2.0]
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'SQLitePCLRaw.batteries_v2.dll' in any of the provided
search directories. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.csh
arp.workspaces\3.11.0\microsoft.codeanalysis.csharp.workspaces.3.11.0.csproj::TargetFramework=netstandard2.0]
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'Humanizer.dll' in any of the provided search directori
es. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.csharp.workspaces\3
.11.0\microsoft.codeanalysis.csharp.workspaces.3.11.0.csproj::TargetFramework=netstandard2.0]
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'Humanizer.dll' in any of the provided search directori
es. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.csharp.workspaces\3
.11.0\microsoft.codeanalysis.csharp.workspaces.3.11.0.csproj::TargetFramework=netstandard2.0]
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'SQLitePCLRaw.core.dll' in any of the provided search d
irectories. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.workspaces.
common\3.11.0\microsoft.codeanalysis.workspaces.common.3.11.0.csproj::TargetFramework=netstandard2.0]
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'SQLitePCLRaw.batteries_v2.dll' in any of the provided
search directories. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.wor
kspaces.common\3.11.0\microsoft.codeanalysis.workspaces.common.3.11.0.csproj::TargetFramework=netstandard2.0]
C:\Users\mavasani\.nuget\packages\microsoft.dotnet.genapi.task\8.0.100-preview.6.23319.8\build\Microsoft.DotNet.GenAPI.
Task.targets(45,5): warning CP1002: Could not resolve reference 'Humanizer.dll' in any of the provided search directori
es. [C:\source-build-reference-packages\artifacts\bin\PackageSourceGenerator\microsoft.codeanalysis.workspaces.common\3
.11.0\microsoft.codeanalysis.workspaces.common.3.11.0.csproj::TargetFramework=netstandard2.0]
```

File: src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/HardcodedSymmetricAlgorithmKeysSources.cs [hunk 6]
```diff
@@ -54,7 +54,7 @@ private static ImmutableHashSet<SourceInfo> BuildSources()
                         (methodName, arguments) =>
                             methodName == "GetBytes" &&
                             arguments.Length == 5 &&
-                            arguments[0].Parameter.Type.SpecialType == SpecialType.System_String,
+                            arguments[0].Parameter?.Type.SpecialType == SpecialType.System_String,
```
Comment (sharwell) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/HardcodedSymmetricAlgorithmKeysSources.cs:57]: The length and value check can be combined:

```
arguments is [{ Parameter.Type.SpecialType: SpecialType.System_String }, _, _, _, _]
```

**Comment by mavasani** [comment]
The PR is now green, except for the Source build failure due to missing pre-built packages detection. Opened https://github.com/dotnet/source-build-reference-packages/pull/727 to help with that.

File: eng/Version.Details.xml [hunk 7]
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Dependencies>
   <ProductDependencies>
-    <Dependency Name="Microsoft.CodeAnalysis" Version="3.3.1">
+    <Dependency Name="Microsoft.CodeAnalysis" Version="3.11.0">
       <Uri>https://github.com/dotnet/roslyn</Uri>
-      <Sha>66a912c9463eebe832cf742d2fe8bb2e1a4600ec</Sha>
+      <Sha>ae1fff344d46976624e68ae17164e0607ab68b10</Sha>
     </Dependency>
     <Dependency Name="Microsoft.SourceBuild.Intermediate.source-build-reference-packages" Version="8.0.0-alpha.1.23305.4">
```
Comment (mavasani) [inline eng/Version.Details.xml:8]: @MichaelSimons I presume we would need to updated the package version here and also update the SHA below. Can you please point me to the package feed where this package is published OR point out the new package version to use?

Comment (mavasani) [reply eng/Version.Details.xml:8]: @MichaelSimons @mmitche Can you please help here?

Comment (MichaelSimons) [reply eng/Version.Details.xml:8]: A darc subscription should be defined to automatically update this dependency.  You can use the darc cli to find the latest version info and feed.

``` bash
> darc get-asset --name Microsoft.SourceBuild.Intermediate.source-build-reference-packages
Looking up assets with name 'Microsoft.SourceBuild.Intermediate.source-build-reference-packages' in the last 30 days
Microsoft.SourceBuild.Intermediate.source-build-reference-packages @ 8.0.0-alpha.1.23328.2
Repository:    https://github.com/dotnet/source-build-reference-packages
Branch:        main
Commit:        5740876021887c3a50d3dae34cc125dbab36316f
Build Number:  20230628.2
Date Produced: 2023-06-28 11:31 AM
Build Link:    https://dev.azure.com/dnceng/internal/_build/results?buildId=2211175
BAR Build Id:  183409
Released:      False
Channels:
- .NET Eng - Latest
Locations:
- https://dev.azure.com/dnceng/internal/_apis/build/builds/2211175/artifacts (Container)
- https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json (NugetFeed)

...
```

**Comment by mavasani** [comment]
@MichaelSimons Seems there are more source build related failures. I'll work on fixing these.

```
.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23269.1/tools/SourceBuild/AfterSourceBuild.proj(68,5): error : (NETCORE_ENGINEERING_TELEMETRY=AfterSourceBuild) 2 new pre-builts discovered! Detailed usage report can be found at /mnt/vss/_work/1/s/artifacts/source-build/self/prebuilt-report/baseline-comparison.xml.
See aka.ms/dotnet/prebuilts for guidance on what pre-builts are and how to eliminate them.
Package IDs are:
System.IO.Pipelines.6.0.3
System.Threading.Channels.6.0.0
```

File: eng/Version.Details.xml [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
 <Dependencies>
   <ProductDependencies>
-    <Dependency Name="Microsoft.CodeAnalysis" Version="3.3.1">
+    <Dependency Name="Microsoft.CodeAnalysis" Version="3.11.0">
       <Uri>https://github.com/dotnet/roslyn</Uri>
-      <Sha>66a912c9463eebe832cf742d2fe8bb2e1a4600ec</Sha>
+      <Sha>ae1fff344d46976624e68ae17164e0607ab68b10</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.SourceBuild.Intermediate.source-build-reference-packages" Version="8.0.0-alpha.1.23305.4">
+    <Dependency Name="Microsoft.SourceBuild.Intermediate.source-build-reference-packages" Version="8.0.0-alpha.1.23328.2">
```
Comment (mavasani) [inline eng/Version.Details.xml:8]:
```suggestion
    <Dependency Name="Microsoft.SourceBuild.Intermediate.source-build-reference-packages" Version="8.0.0-alpha.1.23330.2">
```

File: eng/Version.Details.xml [hunk 9]
```diff
@@ -1,13 +1,13 @@
...
       <Uri>https://github.com/dotnet/source-build-reference-packages</Uri>
-      <Sha>540ccf1a6831993d8ac8bf10bdfd3fde93ecee2e</Sha>
+      <Sha>5740876021887c3a50d3dae34cc125dbab36316f</Sha>
```
Comment (mavasani) [inline eng/Version.Details.xml:10]:
```suggestion
      <Sha>c7e229b7e8cd71c8479e236ae1efff3ad1d740f9</Sha>
```

**Pull request merged**